### PR TITLE
feat: Transactions for crashes

### PR DESCRIPTION
**Pull request by philipphofmann**
## :scroll: Description

Finishes the transaction bound to the scope when the app crashes and stores it to disk. This experimental feature is disabled by default. You can enable it via the option `enableTracingWhenCrashing`.

I'm sorry about the large diff, but the PR only makes sense in one significant change. The largest part is the tests.

Docs PR: https://github.com/getsentry/sentry-docs/pull/11732.

## :bulb: Motivation and Context

Fixes GH-2306

## :green_heart: How did you test it?
Unit tests, simulator, and a release build via TestFlight.

CleanShot 2024-11-05 at 10 52 37@2x

https://sentry-sdks.sentry.io/issues/6044162827/events/4e25fdce217445b5ab3ecd6eb906a77e/?node=txn-efa80d62c9a4472fa8cfef87320638d2&project=5428557

## :pencil: Checklist

You have to check all boxes before merging:

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [x] I updated the docs if needed.
- [x] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

**Comment by philipphofmann** [comment]
@brustolin, I still need to fix some problems with the context before you can review this.

File: Sources/Sentry/SentryTracer.m [hunk 1]
```diff
@@ -521,16 +538,61 @@ - (BOOL)hasUnfinishedChildSpansToWaitFor
...
+            startTimestamp = [SentryDependencyContainer.sharedInstance.dateProvider date];
+        }
+
+        sentry_captureTransactionWithProfile(
+            self.hub, self.dispatchQueue, transaction, startTimestamp);
+        return;
+    }
+#endif // SENTRY_TARGET_PROFILING_SUPPORTED
+
+    [_hub captureTransaction:transaction withScope:_hub.scope];
```
Comment (philipphofmann) [inline Sources/Sentry/SentryTracer.m:573]: This code was already here in a previous version. I only moved it to a different method.

File: Sources/Sentry/SentryScope.m [hunk 2]
```diff
@@ -97,7 +97,7 @@ - (instancetype)initWithScope:(SentryScope *)scope
         [_fingerprintArray addObjectsFromArray:[scope fingerprints]];
         [_attachmentArray addObjectsFromArray:[scope attachments]];
 
-        self.propagationContext = [[SentryPropagationContext alloc] init];
+        self.propagationContext = scope.propagationContext;
```
Comment (philipphofmann) [inline Sources/Sentry/SentryScope.m:100]: I will open an extra PR for this change before merging this PR, as it's a small bug fix that deserves it's own PR.

**Comment by philipphofmann** [comment]
@brustolin, this is ready for review now.

**Review by brustolin — APPROVED** [review]
Thanks for this. PR looks solid. I have a few non blocking comments.

The best approach would be to synchronize between Objective-C and C as transactions happens, but it would be challenging to make that implementation, Im happy with this implementation.

Also, this only finishes the transaction in the scope, but if the user starts a few transactions in parallel without adding it to the scope, those transactions don’t get finished. I think it would be possible to solve this with a message dispatcher, or the prior suggestion.

**Comment by philipphofmann** [comment]
> The best approach would be to synchronize between Objective-C and C as transactions happens

For being a good citizen when we're crashing yes, but it would also mean some of overhead synchronizing all the transaction data consistently to C.

> Also, this only finishes the transaction in the scope, but if the user starts a few transactions in parallel without adding it to the scope, those transactions don’t get finished. I think it would be possible to solve this with a message dispatcher, or the prior suggestion.

If this solution works appropriately, and people ask for it, we cant try to capture all transactions when we're crashing. I think this is good enough for the first iteration.

File: Sources/Sentry/SentryScope.m [hunk 3]
```diff
@@ -601,6 +611,18 @@ - (void)addObserver:(id<SentryScopeObserver>)observer
...
+/**
+ * Make sure to call this inside @c  synchronized(_spanLock) caus this method isn't thread safe.
+ */
+- (NSDictionary *)buildTraceContext:(nullable id<SentrySpan>)span
+{
+    if (span != nil) {
+        return [span serialize];
+    } else {
+        return [self.propagationContext traceContextForEvent];
+    }
```
[the remark this answers is not in the document]
Comment (philipphofmann) [reply Sources/Sentry/SentryScope.m:623]: I don't like ?: so much. I'd like to keep it as is.

File: Sources/Sentry/Public/SentryOptions.h [hunk 4]
```diff
@@ -277,6 +277,17 @@ NS_SWIFT_NAME(Options)
...
+/**
+ * @warning This is an experimental feature and may still have bugs.
+ *
+ * When enabled, the SDK finishes the ongoing transaction bound to the scope and links them to the
+ * crash event when your app crashes. The SDK skips adding profiles to increase the chance of
+ * keeping the transaction.
+ *
+ * @note The default is @c NO .
+ */
+@property (nonatomic, assign) BOOL enableTracingWhenCrashing;
```
[the remark this answers is not in the document]
Comment (philipphofmann) [reply Sources/Sentry/Public/SentryOptions.h]: Hmm, I think we should move all of them there at once. Now, we have two different approaches. I want to keep it for now.

File: Sources/Sentry/SentryCrashIntegration.m [hunk 5]
```diff
@@ -34,6 +36,16 @@
...
 static NSString *const LOCALE_KEY = @"locale";
 
+void
+sentry_finishAndSaveTransaction(void)
+{
+    id<SentrySpan> span = SentrySDK.currentHub.scope.span;
+    if (span != nil && [span isKindOfClass:[SentryTracer class]]) {
+        SentryTracer *tracer = (SentryTracer *)span;
+        [tracer finishForCrash];
+    }
```
[the remark this answers is not in the document]
Comment (philipphofmann) [reply Sources/Sentry/SentryCrashIntegration.m:46]: That's a great idea.

File: Sources/Sentry/SentryScope.m (hunk 2 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryScope.m:100]: Extra PR is here: https://github.com/getsentry/sentry-cocoa/pull/4518. I will wait to merge this PR before merging this one.

**Pull request merged**